Feat: API Playground

oliverqx opened the pull request:
/claim #10283

FelixMalfait commented:
Looks like great progress and clean code 🙌

oliverqx commented:
@FelixMalfait thx! still missing storybook tests, will push in a bit

FelixMalfait commented:
Marking as ready for review so you can see the failing lint tests

FelixMalfait commented:
Actually I went a bit quick, it seemed well structured but nothing loads on my side and tons of linter errors. Does it work on your side? Also as mentioned in the other comment we should switch to Scalar. I don't think we should reuse the playground code from the website since it's different now

FelixMalfait commented:
I resolved some irrelevant Greptile comments related to translations but other comments are probably relevant

oliverqx commented:
@FelixMalfait thanks for reviewing. I did think using the same code wasnt going to be possible, but my plan was to first migrate all of the code from twenty-website to twenty-front and then clean it up, which i was able to do till earlier today

I also addressed all Greptile and fixed linting issues. I did end up using scalar for the playground and also removed a lot of the old implementation, specially the token form and playground wrapped. I think the flow is much cleaner and simple now

oliverqx commented:
almost forgot, I realized this image the [playground cover in light theme](https://github.com/twentyhq/twenty/blob/70c75345e2d2f4831ab3de3dd5e8ae499860004b/packages/twenty-front/src/modules/settings/playground/assets/cover-light.png) has a border set in figma. I asked for permission to edit the figma file, because it looks like theres a double border once it renders in the app

oliverqx commented:
in terms of storybook testing, I've built interaction unit testing, almost like integration UI testing, but I feel like playgrounds themselves (graphiql, scalar) would be out of scope. Please let me know if I should test them

FelixMalfait commented:
Great @oliverqx thanks a lot!!! I'll do a full review tonight

FelixMalfait commented:
Yes the code in twenty-website was ugly, we don't apply the same level of standard to twenty-website (marketing website) as we do in twenty-front. 

I'm sorry I should have told you upfront but one thing I had mentioned over Discord to the other PR was that I don't think the "dropdown to select your API key" works. Because we're not able to access an API Key after it's been generated. So the best we can do is ask the person to paste the key in an input field (but a random api key as placeholder). Eventually if they generated a key in the same session we can prefill it with the `apiKeyTokenFamilyState`(bonus, not required!) which we use to persist it during a session. But if you refresh the page it's lost.

Would you mind also checkin the CI?

Thanks again for this great work!

FelixMalfait commented:
I awarded the bounty given the considerable amount of work you did. But would appreciate if you get this to the end of course! Feel free to ping me on Discord if you want to discuss other tasks in the future (username: ifelx)

FelixMalfait commented on packages/twenty-ui/src/display/icon/components/TablerIcons.ts:
```diff
@@ -40,10 +31,7 @@ export {
-  IconChevronUp,
-  IconChevronsRight,
-  IconCircleDot,
+  IconChevronRight, IconChevronsRight, IconChevronUp, IconCircleDot,
```
Setup your linter on VSCode, this should usually be tackled automatically :)

FelixMalfait commented on packages/twenty-front/src/pages/settings/developers/api-keys/SettingsApiKeys.tsx:
```diff
@@ -0,0 +1,74 @@
+    >
+      <SettingsPageContainer>
+        <StyledContainer>
+          <Section>
```
Make sure you follow exact padding/gaps from Figma, I see quite a few differences :)

FelixMalfait commented on packages/twenty-front/src/pages/settings/developers/api-keys/SettingsApiKeys.tsx:
```diff
@@ -0,0 +1,74 @@
+              title={t`Playground`}
+              description={t`Try our REST or GraphQL API playgrounds.`}
+            />
+            <StyledSettingsApiPlaygroundCoverImage />
```
<img width="512" alt="Light (2)" src="https://github.com/user-attachments/assets/f7cc76ff-b848-40ac-9965-33bc7e5140a3" />

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/assets/cover-dark.png:
<img width="512" alt="Dark" src="https://github.com/user-attachments/assets/9c09b7fa-14b5-47b0-92ac-962c6a0ae3f2" />

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/GraphQLPlayground.tsx:
```diff
@@ -0,0 +1,74 @@
+import { getSettingsPath } from '~/utils/navigation/getSettingsPath';
+
+const SchemaToPath = {
+  [PlaygroundSchemas.CORE.toLocaleLowerCase()]: 'graphql',
```
Rather this no?
```
const SchemaToPath = {
  [PlaygroundSchemas.CORE]: 'graphql
   [PlaygroundSchemas.METADATA]: 'metadata',
};
```
 I would say it's either the map OR the tolowercase but not both (I prefer the map as you did)

oliverqx replied:
yes! that was a quick fix I had to address, cause I wanted to use the enum for both label and value in select options. 

I think that by using translations that can be addressed, but what do you think? I noticed translations are pretty explicit, containing all capital letters that are present in english. And also im not sure if Core and Metadata have a translation at all, kinda like REST and GraphQL

I could also just create a utility function capitalize(string).

FelixMalfait replied:
Ahh I get it now. I think it's best to use "PlaygroundSchemas.XXX" everywhere (so as the key here, not .XXX.lowercase), and yes in the label you can probably use `import { capitalize } from 'twenty-shared';`

FelixMalfait replied:
But honestly an array is probably better because on second thought, it's GraphQL not GraphQl or Graphql

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/PlaygroundSetupForm.tsx:
```diff
@@ -0,0 +1,175 @@
+    resolver: zodResolver(playgroundSetupFormSchema),
+  });
+
+  const { records: apiKeys } = useFindManyRecords<ApiKey>({
```
As discussed, the API key is a JWT which you can't really access again. So we should ask the user to enter the JWT again in an input. This will not work

oliverqx replied:
yes, sorry I forgot to remove the useFindManyRecords hook and options useMemo, I did update the select to an input.  it slipped from my mind, will clean up today

FelixMalfait commented on packages/twenty-front/src/pages/settings/developers/api-keys/SettingsApiKeys.tsx:
```diff
@@ -0,0 +1,74 @@
+            />
+            <StyledSettingsApiPlaygroundCoverImage />
+            <PlaygroundSetupForm />
+          </Section>
```
There are still padding issues.

For example the gap here is 4px:
<img width="245" alt="Screenshot 2025-02-25 at 13 56 59" src="https://github.com/user-attachments/assets/71edf0e6-cfdf-4073-a50e-09399321be69" />

32px: 
![Uploading Screenshot 2025-02-25 at 13.58.45.png…]()

16px:
<img width="823" alt="Screenshot 2025-02-25 at 13 59 18" src="https://github.com/user-attachments/assets/9e612a68-a2fa-4c62-8878-3c2f5ea2494e" />

FelixMalfait replied:
Ah yes sorry I should have been more precise, I hadn't noticed the issue on the playground!

FelixMalfait commented on packages/twenty-front/src/modules/ui/layout/page/components/DefaultLayout.tsx:
```diff
@@ -69,6 +70,7 @@ export const DefaultLayout = () => {
   const theme = useTheme();
   const windowsWidth = useScreenSize().width;
   const showAuthModal = useShowAuthModal();
+  const showPlayground = useShowPlayground();
```
I think this is too specific to playgrounds to deserve to be in DefaultLayout. Maybe we should do something more generic like useShowFullScreen?

FelixMalfait commented on packages/twenty-front/src/modules/app/components/SettingsRoutes.tsx:
```diff
@@ -349,7 +346,11 @@ export const SettingsRoutes = ({
+      <Route path={SettingsPath.APIs} element={<SettingsApiKeys />} />
+      <Route
+        path={SettingsPath.PlaygroundRouter}
+        element={<PlaygroundRouter />}
```
Let's make sure to lazy load everything in settings and in particular the playground stuff, we don't want to slow down the app for everyone on first load with that

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/PlaygroundSetupForm.tsx:
```diff
@@ -0,0 +1,175 @@
+  };
+
+  const onSubmit = async (values: PlaygroundSetupFormValues) => {
+    sessionStorage.setItem('apiKey', values.apiKey);
```
More explicit naming? Like apiKeyForPlayground

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/PlaygroundSetupForm.tsx:
```diff
@@ -0,0 +1,175 @@
+
+    const openAPIReference = await response.json();
+    if (!openAPIReference.tags) {
+      throw new Error('Invalid Token');
```
Translate error message (t). "Invalid API Key" more than token?

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/RestPlayground.tsx:
```diff
@@ -0,0 +1,64 @@
+      ) !important`};
+  }
+`;
+
```
(Scalar comes with Curl / playground so we want that to be prefilled)

oliverqx replied:
oh yes i forgot

FelixMalfait commented on package.json:
```diff
@@ -355,6 +356,7 @@
   "name": "twenty",
   "packageManager": "yarn@4.4.0",
   "resolutions": {
+    "@codemirror/state": "6.4.1",
```
Just for future reference, what was the conflict? Thanks

FelixMalfait commented on packages/twenty-front/src/pages/settings/developers/playground/PlaygroundRouter.tsx:
```diff
@@ -0,0 +1,20 @@
+
+export const PlaygroundRouter = () => {
+  const { schema, type } = useParams<{
+    schema: PlaygroundSchemas;
```
I didn't manage to load the schema on the graphql console, did it work for you?

![Uploading Screenshot 2025-02-25 at 14.27.24.png…]()

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/PlaygroundPage.tsx:
```diff
@@ -0,0 +1,63 @@
+          variant="secondary"
+          accent="default"
+          onClick={closePlayground}
+          ariaLabel={t`Close playground`}
```
Let's do something a little more generic than playground and more like "Exit Full Screen".
We might want to re-use this in other places, for exemple the code editor in Workflow node would benefit from having a full screen mode (could be an interesting followup bounty for you)

oliverqx commented:
@FelixMalfait I just realized my message yesterday night didn't post on github, I'm not sure why. Sorry about that. But thank you for reviewing again. Will address the changes today

FelixMalfait commented on packages/twenty-front/src/modules/settings/playground/components/PlaygroundSetupForm.tsx:
```diff
@@ -0,0 +1,175 @@
+                Icon: IconApi,
+              },
+              {
+                value: PlaygroundTypes.GRAPH_QL,
```
.GRAPHQL

oliverqx commented:
@FelixMalfait hey man, just wanted to give you a headsup, afer abstracting the usePlaygroundSessionHook some settings playground hook does not work properly. will work on a fix in a bit. push in like a couple of hours and then testing will be okay. 

If you could help me with somethings tho, for storybook GraphQL flow looks like this:

1. GraphQLPlayground loads
2. GraphiQL sends introspection schema
3. ggraphqlMock introspection schema returns valid schema

but that is the happy path. 

I was wondering if there is somewhere in the code base you sould point me to where the graphqlMock query errors, so I can test that too
